cppcache/integration/test/PdxJsonTypeTest.cpp

```diff
@@ -142,4 +146,91 @@ TEST(PdxJsonTypeTest, testTwoConsecutiveGets) {
       std::dynamic_pointer_cast<PdxInstance>(region2->get("simpleObject")));
 }
 
+TEST(PdxJsonTypeTest, testTwoObjectsWithSameFieldsHaveTheSameHash) {
+  Cluster cluster{LocatorCount{1}, ServerCount{1}};
+  cluster.start();
+  cluster.getGfsh()
+    .create()
+    .region()
+    .withName("region")
+    .withType("REPLICATE")
+    .execute();
+
+  auto cache = cluster.createCache();
+
+  auto properties = std::make_shared<Properties>();
+  CacheImpl cacheImpl(&cache, properties, false, false, nullptr);
+
+  PdxTypeRegistry pdxTypeRegistry(&cacheImpl);
+
+  apache::geode::client::PdxType m_pdxType1(pdxTypeRegistry, gemfireJsonClassName,false);
+  apache::geode::client::PdxType m_pdxType2(pdxTypeRegistry, gemfireJsonClassName,false);
+
+  m_pdxType1.addVariableLengthTypeField("bar0", "string", PdxFieldTypes::STRING);
+  m_pdxType1.addVariableLengthTypeField("bar1", "string", PdxFieldTypes::STRING);
+
+  m_pdxType2.addVariableLengthTypeField("bar0", "string", PdxFieldTypes::STRING);
+  m_pdxType2.addVariableLengthTypeField("bar1", "string", PdxFieldTypes::STRING);
+
+  EXPECT_EQ(std::hash<PdxType>(m_pdxType1),std::hash<PdxType>(m_pdxType2));
+}
+
+TEST(PdxJsonTypeTest, testTwoObjectsWithDifferentFieldsHaveDifferentHash) {
+  Cluster cluster{LocatorCount{1}, ServerCount{1}};
+  cluster.start();
+  cluster.getGfsh()
+      .create()
+      .region()
+      .withName("region")
+      .withType("REPLICATE")
+      .execute();
+
+  auto cache = cluster.createCache();
+
+  auto properties = std::make_shared<Properties>();
+  CacheImpl cacheImpl(&cache, properties, false, false, nullptr);
+
+  PdxTypeRegistry pdxTypeRegistry(&cacheImpl);
+
+  apache::geode::client::PdxType m_pdxType1(pdxTypeRegistry, gemfireJsonClassName,false);
+  apache::geode::client::PdxType m_pdxType2(pdxTypeRegistry, gemfireJsonClassName,false);
+
+  m_pdxType1.addVariableLengthTypeField("bar0", "string", PdxFieldTypes::STRING);
+  m_pdxType1.addVariableLengthTypeField("bar1", "string", PdxFieldTypes::STRING);
+
+  m_pdxType2.addVariableLengthTypeField("bar2", "string", PdxFieldTypes::STRING);
+  m_pdxType2.addVariableLengthTypeField("bar3", "string", PdxFieldTypes::STRING);
+
+  EXPECT_NE(std::hash<PdxType>(m_pdxType1),std::hash<PdxType>(m_pdxType2));
```

Comment:
I think you problem might be that `PdxType` wasn't imported into the current namespace like many of the other types. Notice lines 195 and 196 reference it with the fully qualified name and this line does not. Add a `using` statement at line 54 and then you should be good, and you can drop the fully qualified names on 195, 196.

Reply:
Thanks for the suggestion, but it did not solve the problem. I used the fully qualified name in the lines you mention because there is other PdxType class imported used in the namespace.

Reply:
Oh yeah forgot about that test object. It but me a few times too and thought it should be renamed. I assumed you tried the fully qualified name on the line 204 too?

Reply:
Thank to your comment I have realized the assertions are wrong. Thing is that in the assertions I was using `m_pdxType1.hashcode()` instead of `std::hash<PdxType>(m_pdxType1)`, that was a change I did right before adding the commit with the tests to the PR and its causing the compilation to fail in other place different than the linker error I was asking for. I have just added the correct code of the assertions.

Reply:
After commenting out the using statement for PdxTest::PdxType, I found the declaration of std::hash wasn't correct in the code (trying to pass 1 param to ctor).  the following change worked for me on line 183 (the first spot I was hitting a compiler gripe):

-   EXPECT_EQ(std::hash<PdxType>(m_pdxType1),std::hash<PdxType>(m_pdxType2));
+   EXPECT_EQ(std::hash<PdxType>{}(m_pdxType1), std::hash<PdxType>{}(m_pdxType2));

I expect the rest are similar...

Reply:
The first linker error is the following:

```
PdxJsonTypeTest.cpp:166: undefined reference to `apache::geode::client::PdxType::PdxType(apache::geode::client::PdxTypeRegistry&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)'
```

Reply:
> the following change worked for me on line 183 (the first spot I was hitting a compiler gripe):

Nice, its simpler that the code I have just added. Now with the current code, you should see the linker error I was not able to solve:

```
PdxJsonTypeTest.cpp:166: undefined reference to `apache::geode::client::PdxType::PdxType(apache::geode::client::PdxTypeRegistry&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)'
... 
and more linker errors from this point
```

Reply:
And now I think we see why PdxTests::PdxType existed in the first place.  apache::geode::client::PdxType is hidden in the library.  The following fixes linker errors for me, in cppcache/src/PdxType.hpp:
```
- class PdxType
+ class APACHE_GEODE_EXPORT PdxType
```
But now we need to have the more philosophical question about whether or not this is the right thing to do.  If we _are_ going to export _PdxType_, the header needs to move from cppcache/src to cppcache/include/geode.  @pivotal-jbarrett any opinion/suggestion?

Reply:
Thanks! With this change Im finally able to verify that the tests are fine.